File: mobile_app/lib/services/comment_service.dart
import 'package:cloud_firestore/cloud_firestore.dart';

class CommentService {
  static final FirebaseFirestore _firestore = FirebaseFirestore.instance;
  static const String _collectionName = 'articles_comments';

  /// Add a new comment to an article
  static Future<Map<String, dynamic>> addComment({
    required String articleId,
    required int userId,
    required String userName,
    required String userType,
    required String content,
  }) async {
    try {
      print('=== Adding Comment ===');
      print('Article ID: $articleId');
      print('User: $userName ($userType)');
      print('Content: $content');

      final commentData = {
        'articleId': articleId,
        'userId': userId,
        'userName': userName,
        'userType': userType,
        'content': content,
        'createdAt': FieldValue.serverTimestamp(),
        'updatedAt': FieldValue.serverTimestamp(),
        'isDeleted': false,
        'parentCommentId': null, // null for top-level comments
      };

      final docRef = await _firestore
          .collection(_collectionName)
          .add(commentData);

      print('Comment added successfully with ID: ${docRef.id}');

      return {
        'success': true,
        'message': 'Comment added successfully',
        'commentId': docRef.id,
      };
    } catch (e) {
      print('Error adding comment: $e');
      return {'success': false, 'message': 'Failed to add comment: $e'};
    }
  }

  /// Add a reply to an existing comment
  static Future<Map<String, dynamic>> addReply({
    required String articleId,
    required String parentCommentId,
    required int userId,
    required String userName,
    required String userType,
    required String content,
  }) async {
    try {
      print('=== Adding Reply ===');
      print('Article ID: $articleId');
      print('Parent Comment ID: $parentCommentId');
      print('User: $userName ($userType)');
      print('Content: $content');

      final replyData = {
        'articleId': articleId,
        'userId': userId,
        'userName': userName,
        'userType': userType,
        'content': content,
        'createdAt': FieldValue.serverTimestamp(),
        'updatedAt': FieldValue.serverTimestamp(),
        'isDeleted': false,
        'parentCommentId': parentCommentId, // Set parent comment ID for replies
      };

      final docRef = await _firestore
          .collection(_collectionName)
          .add(replyData);

      print('Reply added successfully with ID: ${docRef.id}');

      return {
        'success': true,
        'message': 'Reply added successfully',
        'commentId': docRef.id,
      };
    } catch (e) {
      print('Error adding reply: $e');
      return {'success': false, 'message': 'Failed to add reply: $e'};
    }
  }

  /// Get all comments for an article (top-level only)
  static Stream<List<Map<String, dynamic>>> getCommentsStream(
    String articleId,
  ) {
    print('=== Subscribing to comments for article: $articleId ===');
    return _firestore
        .collection(_collectionName)
        .where('articleId', isEqualTo: articleId)
        .where('parentCommentId', isEqualTo: null) // Only top-level comments
        .where('isDeleted', isEqualTo: false)
        .orderBy('createdAt', descending: true) // Newest first
        .snapshots()
        .map((snapshot) {
          print('Received ${snapshot.docs.length} comments from Firestore');
          for (var doc in snapshot.docs) {
            print(
              'Comment ${doc.id}: parentCommentId = ${doc.data()['parentCommentId']}',
            );
          }
          return snapshot.docs.map((doc) {
            final data = doc.data();
            data['id'] = doc.id; // Add document ID to data
            return data;
          }).toList();
        });
  }

  /// Get replies for a specific comment
  static Stream<List<Map<String, dynamic>>> getRepliesStream(String commentId) {
    print('=== Subscribing to replies for comment: $commentId ===');
    return _firestore
        .collection(_collectionName)
        .where('parentCommentId', isEqualTo: commentId)
        .where('isDeleted', isEqualTo: false)
        .orderBy('createdAt', descending: false) // Oldest first for replies
        .snapshots()
        .map((snapshot) {
          print('Received ${snapshot.docs.length} replies from Firestore');
          return snapshot.docs.map((doc) {
            final data = doc.data();
            data['id'] = doc.id; // Add document ID to data
            return data;
          }).toList();
        });
  }

  /// Get comment count for an article (including replies)
  static Future<int> getCommentCount(String articleId) async {
    try {
      final snapshot = await _firestore
          .collection(_collectionName)
          .where('articleId', isEqualTo: articleId)
          .where('isDeleted', isEqualTo: false)
          .get();

      return snapshot.docs.length;
    } catch (e) {
      print('Error getting comment count: $e');
      return 0;
    }
  }

  /// Soft delete a comment
  static Future<Map<String, dynamic>> deleteComment(String commentId) async {
    try {
      await _firestore.collection(_collectionName).doc(commentId).update({
        'isDeleted': true,
        'updatedAt': FieldValue.serverTimestamp(),
      });

      return {'success': true, 'message': 'Comment deleted successfully'};
    } catch (e) {
      print('Error deleting comment: $e');
      return {'success': false, 'message': 'Failed to delete comment: $e'};
    }
  }

  /// Update a comment (edit)
  static Future<Map<String, dynamic>> updateComment({
    required String commentId,
    required String content,
  }) async {
    try {
      await _firestore.collection(_collectionName).doc(commentId).update({
        'content': content,
        'updatedAt': FieldValue.serverTimestamp(),
      });

      return {'success': true, 'message': 'Comment updated successfully'};
    } catch (e) {
      print('Error updating comment: $e');
      return {'success': false, 'message': 'Failed to update comment: $e'};
    }
  }

  /// Format timestamp for display
  static String formatTimestamp(dynamic timestamp) {
    if (timestamp == null) return 'Just now';

    try {
      DateTime date;
      if (timestamp is Timestamp) {
        date = timestamp.toDate();
      } else if (timestamp is int) {
        date = DateTime.fromMillisecondsSinceEpoch(timestamp);
      } else {
        return 'Just now';
      }

      final now = DateTime.now();
      final difference = now.difference(date);

      if (difference.inMinutes < 1) {
        return 'Just now';
      } else if (difference.inMinutes < 60) {
        return '${difference.inMinutes} ${difference.inMinutes == 1 ? 'minute' : 'minutes'} ago';
      } else if (difference.inHours < 24) {
        return '${difference.inHours} ${difference.inHours == 1 ? 'hour' : 'hours'} ago';
      } else if (difference.inDays < 7) {
        return '${difference.inDays} ${difference.inDays == 1 ? 'day' : 'days'} ago';
      } else {
        return '${date.day}/${date.month}/${date.year}';
      }
    } catch (e) {
      return 'Just now';
    }
  }
}

File: web_portal/src/services/commentService.js
import { getFirestore, collection, addDoc, query, where, orderBy, onSnapshot, Timestamp, serverTimestamp } from 'firebase/firestore';
import { db } from '../config/firebase';

class CommentService {
  constructor() {
    this.collectionName = 'articles_comments';
  }

  /**
   * Add a new comment to an article
   * @param {Object} commentData - Comment data
   * @param {string} commentData.articleId - Article ID
   * @param {number} commentData.userId - User ID
   * @param {string} commentData.userName - User name
   * @param {string} commentData.userType - User type (Volunteer, Caregiver, etc.)
   * @param {string} commentData.content - Comment content
   * @param {string} commentData.parentCommentId - Parent comment ID for replies (optional)
   * @returns {Promise<Object>} Response with success status
   */
  async addComment({ articleId, userId, userName, userType, content, parentCommentId = null }) {
    try {
      if (!db) {
        throw new Error('Firebase not initialized');
      }

      console.log('Adding comment:', { articleId, userId, userName, userType, content, parentCommentId });

      const commentData = {
        articleId,
        userId,
        userName,
        userType,
        content,
        createdAt: serverTimestamp(),
        updatedAt: serverTimestamp(),
        isDeleted: false,
        parentCommentId: parentCommentId || null, // null for top-level comments, ID for replies
      };

      const docRef = await addDoc(collection(db, this.collectionName), commentData);

      console.log('Comment added successfully with ID:', docRef.id);

      return {
        success: true,
        message: parentCommentId ? 'Reply added successfully' : 'Comment added successfully',
        commentId: docRef.id,
      };
    } catch (error) {
      console.error('Error adding comment:', error);
      return {
        success: false,
        message: `Failed to add comment: ${error.message}`,
        error: error.message
      };
    }
  }

  /**
   * Get replies for a specific comment
   * @param {string} parentCommentId - Parent comment ID
   * @param {Function} callback - Callback function to receive replies
   * @returns {Function} Unsubscribe function
   */
  subscribeToReplies(parentCommentId, callback) {
    try {
      if (!db) {
        console.error('Firebase not initialized');
        callback([]);
        return () => {};
      }

      console.log('Subscribing to replies for comment:', parentCommentId);

      const q = query(
        collection(db, this.collectionName),
        where('parentCommentId', '==', parentCommentId),
        where('isDeleted', '==', false)
      );

      const unsubscribe = onSnapshot(q, 
        (snapshot) => {
          console.log('Received', snapshot.docs.length, 'replies from Firestore');
          const replies = snapshot.docs.map((doc) => ({
            id: doc.id,
            ...doc.data()
          }));
          
          // Sort replies by createdAt (oldest first for replies)
          replies.sort((a, b) => {
            const timeA = a.createdAt?.toMillis ? a.createdAt.toMillis() : 0;
            const timeB = b.createdAt?.toMillis ? b.createdAt.toMillis() : 0;
            return timeA - timeB; // Oldest first
          });
          
          callback(replies);
        },
        (error) => {
          console.error('Error subscribing to replies:', error);
          callback([]);
        }
      );

      return unsubscribe;
    } catch (error) {
      console.error('Error setting up reply subscription:', error);
      callback([]);
      return () => {};
    }
  }

  /**
   * Subscribe to comments for an article (real-time updates)
   * @param {string} articleId - Article ID
   * @param {Function} callback - Callback function to receive comments
   * @returns {Function} Unsubscribe function
   */
  subscribeToComments(articleId, callback) {
    try {
      if (!db) {
        console.error('Firebase not initialized');
        callback([]);
        return () => {};
      }

      console.log('Subscribing to comments for article:', articleId);

      // Simpler query - just filter by articleId and isDeleted
      const q = query(
        collection(db, this.collectionName),
        where('articleId', '==', articleId),
        where('isDeleted', '==', false)
      );

      const unsubscribe = onSnapshot(q, 
        (snapshot) => {
          console.log('Received', snapshot.docs.length, 'comments from Firestore');
          const comments = snapshot.docs.map((doc) => ({
            id: doc.id,
            ...doc.data()
          }));
          
          // Filter top-level comments and sort by createdAt on the client side
          const topLevelComments = comments.filter(c => !c.parentCommentId || c.parentCommentId === null);
          
          topLevelComments.sort((a, b) => {
            const timeA = a.createdAt?.toMillis ? a.createdAt.toMillis() : 0;
            const timeB = b.createdAt?.toMillis ? b.createdAt.toMillis() : 0;
            return timeB - timeA; // Newest first
          });
          
          console.log('Filtered top-level comments:', topLevelComments.length);
          callback(topLevelComments);
        },
        (error) => {
          console.error('Error subscribing to comments:', error);
          console.error('Error details:', error.message, error.code);
          callback([]);
        }
      );

      return unsubscribe;
    } catch (error) {
      console.error('Error setting up comment subscription:', error);
      callback([]);
      return () => {};
    }
  }

  /**
   * Get comment count for an article
   * @param {string} articleId - Article ID
   * @returns {Promise<number>} Comment count
   */
  async getCommentCount(articleId) {
    try {
      if (!db) {
        return 0;
      }

      const q = query(
        collection(db, this.collectionName),
        where('articleId', '==', articleId),
        where('isDeleted', '==', false)
      );

      // Note: This is a simplified version
      // For better performance, you might want to maintain a count in the article document
      return 0; // Placeholder - implement count logic as needed
    } catch (error) {
      console.error('Error getting comment count:', error);
      return 0;
    }
  }

  /**
   * Format timestamp for display
   * @param {any} timestamp - Firestore timestamp or Date
   * @returns {string} Formatted time string
   */
  formatTimestamp(timestamp) {
    if (!timestamp) return 'Just now';

    try {
      let date;
      if (timestamp && typeof timestamp.toDate === 'function') {
        // Firestore Timestamp
        date = timestamp.toDate();
      } else if (timestamp instanceof Date) {
        date = timestamp;
      } else if (typeof timestamp === 'number') {
        date = new Date(timestamp);
      } else {
        return 'Just now';
      }

      const now = new Date();
      const diffMs = now - date;
      const diffMins = Math.floor(diffMs / 60000);
      const diffHours = Math.floor(diffMs / 3600000);
      const diffDays = Math.floor(diffMs / 86400000);

      if (diffMins < 1) {
        return 'Just now';
      } else if (diffMins < 60) {
        return `${diffMins} ${diffMins === 1 ? 'minute' : 'minutes'} ago`;
      } else if (diffHours < 24) {
        return `${diffHours} ${diffHours === 1 ? 'hour' : 'hours'} ago`;
      } else if (diffDays < 7) {
        return `${diffDays} ${diffDays === 1 ? 'day' : 'days'} ago`;
      } else {
        return date.toLocaleDateString();
      }
    } catch (error) {
      console.error('Error formatting timestamp:', error);
      return 'Just now';
    }
  }
}

// Create and export a singleton instance
const commentService = new CommentService();
export default commentService;
